File: java/Kalk/scalare.java
package Kalk;

public class scalare implements operando {
	
	private double reale;
	
	/** Costruttore ad un parametro*/
	public scalare(double r){
		this.reale=r;
	}

	/** Costruttore senza parametri*/	
	public scalare(){
		this(0);
	}

		/** Costruttore di copia*/
	public scalare (scalare s){
		this.reale=s.reale;
	}

	/** Funzione che effettua la moltiplicazione
	 * throws ErrCalcolo*/
	@Override
	public operando moltiplicazione(operando pro) throws ErrCalcolo {
		if(pro instanceof scalare) {
			scalare s=(scalare)pro;
			this.reale=this.reale*s.reale;
		}
		else {
    		return pro.moltiplicazione(this);
    	}
		return this;
	}

	/** Funzione che effettua la somma
	 * throws ErrCalcolo*/
	@Override
	public operando somma(operando sum) throws ErrCalcolo {
		if(sum instanceof scalare) {
			scalare s=(scalare)sum;
			this.reale=this.reale+s.reale;
		}
		else {
    		return sum.somma(this);
    	}
		return this;
	}

	/** Funzione che effettua la sottrazione
	 * throws ErrCalcolo*/
	@Override
	public operando sottrazione(operando sot) throws ErrCalcolo {
		if(sot instanceof scalare) {
			scalare s=(scalare)sot;
			this.reale=this.reale-s.reale;
		}
		else {
    		return sot.sottrazione(this);
    	}
		return this;
	}

	/** Funzione di accesso al valore*/
	public double get_val() {
		return reale;
	}

	/** Funzione che confronta lo scalare con un double*/
	public boolean equals(double eq) {
		return eq==reale;
	}

	/** Overload del metodo toString()*/
	public String toString() {
		return reale+" ";
	}
}

File: scalare.cpp
#include "scalare.h"

// costruttore a 0 o 1 parametro
scalare::scalare(const double & i):reale(i)
{}

// costruttore di copia
scalare::scalare(const scalare& s):reale(s.get_val()){}

// funzione per accedere al valore
double scalare::get_val() const{
    return reale;
}

// funzione per assegnare il valore
void scalare::set_val(const double & i){
    reale=i;
}

// operatore di moltiplicazione ereditato dalla classe base
operando* scalare::operator*(operando const * i){
    const scalare *a=dynamic_cast<const scalare*>(i);
    if(a!=0){
        reale*=a->get_val();
    }
    else{
        throw ecc_math("operazione non implementata");
    }
    return this;
}

// operatore di somma ereditato dalla classe base
operando* scalare::operator+(operando const * i){
    const scalare *a=dynamic_cast<const scalare*>(i);
    if(a!=0){
        reale+=a->get_val();
    }
    else{
        throw ecc_math("operazione non implementata");
    }
    return this;
}

// operatore di sottrazione ereditato dalla classe base
operando* scalare::operator-(operando const * i){
    const scalare *a=dynamic_cast<const scalare*>(i);
    if(a!=0){
        reale-=a->get_val();
    }
    else{
        throw ecc_math("operazione non implementata");
    }
    return this;
}

// funzione che converte l'oggetto in una stringa per poterla stampare a video
std::string scalare::toString() const{
    return std::to_string(reale);
}

// funzione di uguaglianza con un double
bool scalare::equals(const double & c){
    return reale==c;
}

// funzione accessoria per la stampa
void scalare::print(std::ostream & os) const{
    os<<get_val();
}

// operatore di output
std::ostream& operator<<(std::ostream& os, const scalare& obj){
    obj.print(os);
    return os;
}

// operatore di input
std::istream& operator>>(std::istream& is, scalare& obj){
    std::string ca;
    double real=0;
    is>>real;
    obj.set_val(real);
    return is;
}

